### ISSUE TITLE
Possible bug when building pag for global struct ?

### ISSUE THREAD

**Issue by yiyuaner** [SVF-tools/SVF#87]
In function PAGBuilder::InitialGlobal, we have the following code fragment:
```
else if (isa<ConstantStruct>(C)) {
    const StructType *sty = cast<StructType>(C->getType());
    const std::vector<u32_t>& offsetvect =
        SymbolTableInfo::Symbolnfo()->getStructOffsetVec(sty);
    for (u32_t i = 0, e = C->getNumOperands(); i != e; i++) {
        u32_t off = offsetvect[i];
        InitialGlobal(gvar, cast<Constant>(C->getOperand(i)), offset + off, i);
    }
} 

```

if we have gvar pointing to a nested structure like:
```
struct A {
     double a
     struct B {
         int b;
         int c;
     } b;
     float d;
}
```
and with the initializer C (psudocode):
```
{1.0, {2, 3}, 4.0}
```
then we will get the following calls:
```
InitialGlobal(gvar, Constant(1.0), 0, 0);
InitialGlobal(gvar, ConstantStruct({2,3}), 1, 1);
InitialGlobal(gvar, Constant(4.0), 2, 2);
```
the second call above for initializing inner struct B will invoke another call:
```
InitialGlobal(gvar, Constant(2), 1+0, 0); // here we pass 0 as fieldidx param
```
when we try to get the field, we will invoke the following calls:
```
NodeID field = getGlobalVarField(gvar, 1, 0); 
pag->getGepValNode(gvar, LocationSet(1), gvartype, 0); // here gvartype is type for struct A.
const std::vector<FieldInfo> &fieldinfo = symInfo->getFlattenFieldInfoVec(gvartype); // FieldInfoVec for struct A
const Type *type = fieldinfo[0].getFlattenElemTy();
// index into A's FieldInfoVec, get type double, wrong! we should get type int instead.
```
The fieldidx param of getGlobalVarField is the index for the inner struct B but we use the fieldinfo vec of
outter struct A. So we get the wrong type for the gepNode.

Is this a bug ?

**Comment by yiyuaner** [comment]
@yuleisui 
I test with the latest version of SVF, and some other issues occured.
With the following test code:
```
struct A {
     double *a;
     struct B {
         int *b;
         int *c;
     } b;
     float *d;
};

double a;
int b,c;
float d;

A obj = {&a, {&b, &c}, &d};

int main() {
    A obj;
    return 0;
}
```
The corresponding llvm disassembled  bitcode is:
```
%struct.A = type { double*, %"struct.A::B", float* }
%"struct.A::B" = type { i32*, i32* }

@a = global double 0.000000e+00, align 8
@b = global i32 0, align 4
@c = global i32 0, align 4
@d = global float 0.000000e+00, align 4
@obj = global %struct.A { double* @a, %"struct.A::B" { i32* @b, i32* @c }, float* @d }, align 8

; Function Attrs: noinline norecurse nounwind optnone sspstrong uwtable
define i32 @main() #0 {
  %1 = alloca i32, align 4
  %2 = alloca %struct.A, align 8
  store i32 0, i32* %1, align 4
  ret i32 0
}
```
I test it with  wpa -ander -dump-pag test.bc, the PAG doesn't seem right:

![chart](https://user-images.githubusercontent.com/20484362/44128885-ed92325c-a034-11e8-9af4-d925f36cd73e.png)

All store edge from node a,b,c,d points to the same field: gep node 24. We need different gep node for different field in the struct, right? 

I have tested it with an earlier version of SVF, these store edges will point to different gep nodes.
I think it has something to do with the change of handling GepValNodeMap.

**Comment by yuleisui** [comment]
Yes, we are fixing the last commit of gepvalnode now.

**Comment by yiyuaner** [comment]
For the same testing code input (repeated here):
```
struct A {
     double *a;
     struct B {
         int *b;
         int *c;
     } b;
     float *d;
};

double a;
int b,c;
float d;

A obj = {&a, {&b, &c}, &d};

int main() {
    A obj;
    return 0;
}
```
llvm bc:
```
%struct.A = type { double*, %"struct.A::B", float* }
%"struct.A::B" = type { i32*, i32* }

@a = global double 0.000000e+00, align 8
@b = global i32 0, align 4
@c = global i32 0, align 4
@d = global float 0.000000e+00, align 4
@obj = global %struct.A { double* @a, %"struct.A::B" { i32* @b, i32* @c }, float* @d }, align 8

; Function Attrs: noinline norecurse nounwind optnone sspstrong uwtable
define i32 @main() #0 {
  %1 = alloca i32, align 4
  %2 = alloca %struct.A, align 8
  store i32 0, i32* %1, align 4
  ret i32 0
}

```
I add some code to PointerAnalysis::dumpAllTypes():
```
void PointerAnalysis::dumpAllTypes()
{
    for (NodeBS::iterator nIter = this->getAllValidPtrs().begin();
            nIter != this->getAllValidPtrs().end(); ++nIter) {
        const PAGNode* node = getPAG()->getPAGNode(*nIter);
        if (isa<DummyObjPN>(node) || isa<DummyValPN>(node))
            continue;

        outs() << "##<" << node->getValue()->getName() << "> ";
        outs() << "Source Loc: " << getSourceLoc(node->getValue());
        outs() << "\nNodeID " << node->getId() << "\n";

        // code added
        if (isa<GepValPN>(node)) {
            const llvm::Type* gepNodeType = cast<GepValPN>(node)->getType();
            outs() << "llvm Type for gep node:\n";
            gepNodeType->print(outs());
        }
        ......
```
Run the command: wpa -ander -print-pag -print-type test.bc and wpa -ander -dump-pag test.bc.
The PAG is correct:
![chart](https://user-images.githubusercontent.com/20484362/44136296-cf7fe468-a05b-11e8-8e06-719f77fc5d9c.png)

We can see from the pag that b is stored into gep node 25, which should have llvm type i32*. But the
output of dumpAllTypes shows that it has type double*:
```
##<obj> Source Loc: Glob
NodeID 25
llvm Type for gep node:
double* 
```
I have tried to fix it in the pull request: #89 and tested it. It seems to work.
